### PR TITLE
Plugin change for EHSN 2.1.0

### PR DESCRIPTION
**Pull request by yanxuYX**
Hey Doug,

Here are some changes to the plugin:
- Apply GC on WaterLevel
- Update Version Number
- Handle Engineered structure and Other Methods
- Ignore RP and TP in WLHeader
- Add reading type mapping for Water Level
- Set reading type to blank in "AddReading"
- Include logger time from MeasResults and time in level check table rows in "ExtractAllTimes"

Thanks,
Xu

### CODE REVIEW

**Comment by TheRealDougSchmidt** [comment]
@yanxuYX I'll need to take a look at this a bit later next week. I'm a bit swamped on other project deliverables at the moment.

But I've just granted you Write access to the repo.

So if you want to click the "Merge pull request" button, you can certainly do that now. (So I won't any longer be the bottleneck to deploying a required WSC fix).

I'll still do the review, but that won't happen until later next week.

**Comment by yanxuYX** [comment]
Hey @DougSchmidt-AI 
Thanks for letting me know, I will merge the changes now and refactor the code if it needs to after your review. It seems that I don't have write access to merge.

image

File: src/EhsnPlugin/DataModel/ParsedEhsnLevelSurvey.cs [hunk 1]
```diff
@@ -35,14 +35,16 @@ public class ParsedEhsnLevelSurvey
...
         {
-            if (value == null || !value.Contains(PrimaryPrefix))
-                return value;
-
-            return value.Replace("*", string.Empty).Trim();
+            if (value.StartsWith(PrimaryPrefix))
+                return value.Substring(PrimaryPrefix.Length).Trim();
+            if (value.EndsWith(PrimarySuffix))
+                return value.TrimEnd('*').Trim();
+            return value;
```
Comment (TheRealDougSchmidt) [inline src/EhsnPlugin/DataModel/ParsedEhsnLevelSurvey.cs:47]: @yanxuYX I'm finally getting around to the review! (Sorry, I've been swamped).

Not a big deal at all, but if you just need to remove leading and trailing asterisks from the benchnames, then you can do that with a single line in C#, using the null-aware method call (the `?.` operator):

```C#
public static string SanitizeBenchmarkName(string value)
{
    return value?.Trim('*');
}
```

File: src/EhsnPlugin/Mappers/ReadingMapper.cs [hunk 2]
```diff
@@ -203,62 +203,79 @@ private void AddGageHeightReading(List<Reading> readings, EHSNStageMeasStageMeas
...
-                    : ReadingType.Unknown;
+            if (string.IsNullOrWhiteSpace(readingType))
+            {
+                reading.ReadingType = ReadingType.Unknown;
+            }
+            else
+            {
+                string formattedReadingType = readingType.Replace(" ", String.Empty);
+                reading.ReadingType = (ReadingType)Enum.Parse(typeof(ReadingType), formattedReadingType);
+            }
```
Comment (TheRealDougSchmidt) [inline src/EhsnPlugin/Mappers/ReadingMapper.cs:245]: You can also use the Enum.TryParse<TEnum> method for more robust enum parsing.

This whole if-then-else block becomes this line.

```C#
reading.ReadingType = Enum.TryParse<ReadingType>(readingType, true, out var value) ? value : ReadingType.Unknown;
```
That will trim any leading/trailing whitespace, parse the enum value case-insensitively and fallback to Unknown when the parse fails.

File: src/EhsnPlugin/Mappers/ReadingMapper.cs [hunk 3]
```diff
@@ -270,17 +287,20 @@ private void AddReadingComments(Reading reading, double? src, string srcAction,
...
+                return value;
+            if (invalidBM.Contains(value))
                 return null;
-
-            return ParsedEhsnLevelSurvey.SanitizeBenchmarkName(levelCheck.reference);
+            if (value.StartsWith(PrimaryPrefix))
+                return value.Substring(PrimaryPrefix.Length).Trim();
+            if (value.EndsWith(PrimarySuffix))
+                return value.TrimEnd('*').Trim();
+            return value;
```
Comment (TheRealDougSchmidt) [inline src/EhsnPlugin/Mappers/ReadingMapper.cs:303]: This could benefit from the same `value?.Trim('*')` technique as earlier noted.

**Review by TheRealDougSchmidt — COMMENTED** [review]
@yanxuYX Overall those changes look good.

I only added some guidance to consider for future edits.

**Pull request merged**